CMSgov/mint-app
# Easi 1980/collaborator access control

# EASI-1980

## Changes and Description

This change focuses on disallowing making changes to items that are part of model plan that you aren't a collaborator on. Assessment users are excluded from this and are able to make changes to every item.  Normal users that try to update sections that they aren't a collaborator on will get an error message that says they don't have permission to do that.


- Resolvers pass principal struct instead of string.
- Add SQL to check if user is a collaborator by ModelPlanID
- Add SQL to check if user is a collaborator by PlanDiscussionIOD
- Add Check in pre-update for base -struct
    - Manual addition to structs not able to use base struct update currently
        - Plan Document
        - Plan Collaborator
        - Plan Discussion Reply 
- Add interface for model plan relations to enforce permissions before each update
- Add interface for model plan relations to enforce permissions before delete and create for relations that are manually created or deleted
    - EG, collaborator, document, discussion, and discussion reply.
- Add Embedded struct for most use cases that implements the model plan relation
    - Manually implement the functionality for structs that can't use the embedded struct ( EG Model Plan)
  


## How to test this change

1. Run the app
2. Try to update data on any form section where you aren't a collaborator
   a. I suggest using the front end UI. You can  navigate by URL to a model plan you don't have permission to edit. Try to edit, delete, and add to that model plan
   b. I suggest repeating the activity as a user with assessment as well as the basic user role. Note that you can take actions on all model plans. 


##Other things to review
1. Do you agree with the error logging? 
2. Do you agree with the interface naming and uses?
3. Is this in the most useful format for the front end?

## PR Author Review Checklist



## PR Reviewer Guidelines
- It's best to pull the branch locally and test it, rather than just looking at the code online!
- Check that all code is adequately covered by tests - if it isn't feel free to suggest the addition of tests.
- Always make comments, even if it's minor, or if you don't understand why something was done.
